seqfold/fold.go

```diff
@@ -0,0 +1,875 @@
+// Returns:
+//     List[Struct]: A list of structures. Stacks, bulges, hairpins, etc.
+func Fold(seq string, temp float64) []NucleicAcidStructure {
+	//v_cache, w_cache := cache(seq, temp)
+	//n := len(seq)
+	fc, err := NewFoldingContext(seq, temp)
+	if err != nil {
+		fmt.Printf("%+v\n", err)
+		os.Exit(1)
+	}
+
+	// get the minimum free energy structure out of the cache
+	return Traceback(0, len(seq)-1, fc)
+}
+
+// Fold the sequence and return just the delta G of the structure
+//
+// Args:
+//     seq: The sequence to fold
+// Keyword Args:
+//     temp: The temperature to fold at
+// Returns:
+//     float: The minimum free energy of the folded sequence
+func dG(seq string, temp float64) float64 {
+	structs := Fold(seq, temp)
+	sumE := 0.0
+	for _, s := range structs {
+		sumE += s.E
+	}
+	return RoundFloat(sumE, 2)
+}
+
+// Fold a nucleic acid sequence and return the estimated dg of each (i,j) pairing.
+//
+// Args:
+//     seq: The nucleic acid sequence to fold
+// Keyword Args:
+//     temp: The temperature to fold at
+// Returns:
+//     Cache: A 2D matrix where each (i, j) pairing corresponds to the
+//         minimum free energy between i and j
+func dGCache(seq string, temp float64) ([][]float64, error) {
+	//_, w_cache := cache(seq, temp)
+	fc, err := NewFoldingContext(seq, temp)
+	if err != nil {
+		return nil, fmt.Errorf("error: %w", err)
+	}
+	wCache := fc.W
+
+	cache := make([][]float64, len(seq))
+	for i := range cache {
+		cache[i] = make([]float64, len(seq))
+	}
+
+	for i, wrow := range wCache {
+		for j, s := range wrow {
+			cache[i][j] = s.E
+		}
+	}
+
+	return cache, nil
+}
+
+// Get the dot bracket notation for a secondary structure.
+//
+// Args:
+//     structs: A list of structs, usually from the fold function
+// Returns:
+//     str: The dot bracket notation of the secondary structure
+func DotBracket(structs []NucleicAcidStructure) string {
+	// log structure with dot-bracket notation
+	// result = ["."] * (max(j for s in structs for _, j in s.ij) + 1)
+	maxj := 0
+	for _, s := range structs {
+		for _, ij := range s.Inner {
+			if ij.J > maxj {
+				maxj = ij.J
+			}
+		}
+	}
+	maxj += 1
+	result := make([]byte, maxj)
+	for i := range result {
+		result[i] = '.'
+	}
+	for _, s := range structs {
+		if len(s.Inner) == 1 {
+			ij := s.Inner[0]
+			result[ij.I] = '('
+			result[ij.J] = ')'
+		}
+	}
+	return string(result)
+}
+
+// Find and return the lowest free energy structure in Sij subsequence
+//
+// Figure 2B in Zuker and Stiegler, 1981
+// Args:
+//     seq: The sequence being folded
+//     i: The start index
+//     j: The end index (inclusive)
+//     temp: The temperature in Kelvin
+//     v_cache: Free energy cache for if i and j bp
+//     w_cache: Free energy cache for lowest energy structure from i to j. 0 otherwise
+// Returns:
+//     float: The free energy for the subsequence from i to j
+func W(i, j int, fc FoldContext) NucleicAcidStructure {
```

Comment:
I forget what `W` stands for and we should probably rename it to something more readable. Same goes fo `V`

Reply:
these single letter variables from the original paper + seqfold implementation are funkayy. Should find a way to rename them.